keps/sig-api-machinery/20200123-client-go-ctx.md

```diff
@@ -0,0 +1,188 @@
+participating-sigs:
+- sig-api-machinery
+approvers:
+- "@deads2k"
+- "@liggitt"
+- "@lavalamp"
+creation-date: 2020-01-23
+last-updated: 2020-01-23
+status: provisional
+---
+
+# Context support in k8s.io/client-go
+
+## Table of Contents
+
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+- [Proposal](#proposal)
+- [Design Details](#design-details)
+  - [Risks and Mitigations](#risks-and-mitigations)
+  - [Graduation Criteria](#graduation-criteria)
+- [Alternatives](#alternatives)
+<!-- /toc -->
+
+## Summary
+
+This proposal outlines a refactoring that would allow propagating of
+[context](https://golang.org/pkg/context/) with
+[client-go](https://github.com/kubernetes/client-go) with API calls made using
+k8s.io/client-go clientsets.
+
+## Motivation
+
+When using client-go, external calls to the Kubernetes API are made in order to
+manage resources. Adding support for context propagation along these request
+paths enables clean and efficient implementation of cross cutting functionality.
+This functionality initially will include:
+
+- Support for request timeout and cancelation that frees the calling thread. Due
+  to issues with authentication and authorization performance, this was
+  something that was
+  [retrofited](https://github.com/kubernetes/kubernetes/pull/83064) in various
+  auth\* expansions but is generally useful.
+- Support for distributed tracing. Active
+  [efforts](https://github.com/kubernetes/enhancements/blob/master/keps/sig-instrumentation/0034-distributed-tracing-kep.md#propagating-context-through-objects)
+  by SIG Instrumentation to support distributed tracing are either blocked or
+  would be substantially simplified with this enhancement.
+
+### Goals
+
+- Allow consumers of k8s.io/client-go to propage context with API requests made
+  using a clientset.
+
+### Non-Goals
+
+- Plumb context to client-go callsites in core Kubernetes. We believe that by
+  making this feature available, this type of refactoring will happen over time.
+
+## Proposal
+
+The recommended approach is to pass the context with each action on a resource.
+We will modify the signatures of all client interfaces to accept a
+`context.Context` as the first argument.  This is idiomatic, explicit, and
+produces the least error prone API.
+
+## Design Details
+
+This is an example change to `Pods.Get` implementation:
+
+```go 
+func (c *pods) Get(ctx context.Context, name string, options meta_v1.GetOptions) (result *v1.Pod, err error) {
+	Result = &v1.Pod{}
+	err = c.client.Get().
+		Context(ctx).
+		Namespace(c.ns).
+		Resource("pods").
+		Name(name).
+		VersionedParams(&options, scheme.ParameterCodec).
+		Do().
+		Into(result)
+	return
+}
+```
+
+Here we extend the signature to add `context.Context` as the first argument,
+then attach the context to the request.
+
+Initially `lister` and `informer` APIs (both generated and generic) will not
+change.
```

Comment:
I'd suggest dropping this line... I don't anticipate those changing as they have a lifetime longer than one request